src/uu/stat/src/stat.rs

```diff
@@ -26,57 +26,61 @@ use std::os::unix::prelude::OsStrExt;
 use std::path::Path;
 use std::{cmp, fs, iter};
 
-macro_rules! check_bound {
-    ($str: ident, $bound:expr, $beg: expr, $end: expr) => {
-        if $end >= $bound {
-            return Err(USimpleError::new(
-                1,
-                format!("{}: invalid directive", $str[$beg..$end].quote()),
-            ));
-        }
-    };
+fn check_bound(slice: &str, bound: usize, beg: usize, end: usize) -> Result<(), Box<dyn UError>> {
+    if end >= bound {
+        return Err(USimpleError::new(
+            1,
+            format!("{}: invalid directive", slice[beg..end].quote()),
+        ));
+    }
+    Ok(())
 }
-macro_rules! fill_string {
-    ($str: ident, $c: expr, $cnt: expr) => {
-        iter::repeat($c)
-            .take($cnt)
-            .map(|c| $str.push(c))
-            .all(|_| true)
-    };
+
+fn fill_string(strs: &mut String, c: char, cnt: usize) {
+    iter::repeat(c)
+        .take(cnt)
+        .map(|c| strs.push(c))
+        .all(|_| true);
 }
-macro_rules! extend_digits {
-    ($str: expr, $min: expr) => {
-        if $min > $str.len() {
-            let mut pad = String::with_capacity($min);
-            fill_string!(pad, '0', $min - $str.len());
-            pad.push_str($str);
-            pad.into()
-        } else {
-            $str.into()
-        }
-    };
+
+fn extend_digits(strs: &str, min: usize) -> Cow<'_, str> {
+    if min > strs.len() {
+        let mut pad = String::with_capacity(min);
+        fill_string(&mut pad, '0', min - strs.len());
+        pad.push_str(strs);
+        return pad.into();
+    } else {
+        return strs.into();
+    }
 }
-macro_rules! pad_and_print {
-    ($result: ident, $str: ident, $left: expr, $width: expr, $padding: expr) => {
-        if $str.len() < $width {
-            if $left {
-                $result.push_str($str.as_ref());
-                fill_string!($result, $padding, $width - $str.len());
-            } else {
-                fill_string!($result, $padding, $width - $str.len());
-                $result.push_str($str.as_ref());
-            }
+
+fn pad_and_print<S: AsRef<str>>(
```

Comment:
I think this can be simplified by using format specifiers in the `print!` call. The only problem is that we have to make it slightly less general and only support `'0'` and `' '` as padding characters, because the format specification does not support general fill characters. It would look something like:
```rust
enum Padding {
    Zero,
    Space,
}

fn pad_and_print(result: &str, left: bool, width: usize, padding: Padding) {
    match (left, padding) {
        (false, Padding::Zero) => print!("{result:0>width$}"),
        (false, Padding::Space) => print!("{result:>width$}"),
        (true, Padding::Zero) => print!("{result:0<width$}"),
        (true, Padding::Space) => print!("{result:<width$}"),
    };
}
```

I think that with this implementation we can get rid of `fill_string` altogether.

Reply:
Thanks for the input, I've updated the changes. Please take a look at it at your convenience. Thanks :)